File: src/sam.cpp
#include "sam.hpp"
#include <algorithm>
#include <ostream>
#include <sstream>
#include <iostream>

#define SAM_UNMAPPED_MAPQ 0
#define SAM_UNMAPPED_MAPQ_STRING "0"

using namespace klibpp;

/*
 * SAM columns
 *
 * 1 QNAME query template name
 * 2 FLAG  bitwise flag
 * 3 RNAME reference sequence name
 * 4 POS   1-based leftmost mapping position (set to 0 for unmapped read w/o pos)
 * 5 MAPQ  mapping quality (255 if unavailable)
 * 6 CIGAR
 * 7 RNEXT reference name of mate/next read
 * 8 PNEXT position of mate/next read
 * 9 TLEN template length
 * 10 SEQ
 * 11 QUAL
 * 12 optional fields
*/

/* Strip the /1 or /2 suffix from a read name */
std::string strip_suffix(const std::string& name) {
    auto len = name.length();
    if (
        len >= 2
        && name[len - 2] == '/'
        && (name[len - 1] == '1' || name[len - 1] == '2')
    ) {
        return std::string{name, 0, len - 2};
    } else {
        return name;
    }
}

void Sam::append_rg() {
    sam_string.append(tail);
}

void Sam::append_details(const Details& details) {
    std::stringstream s;
    s << "\tna:i:" << details.nams
        << "\tnr:i:" << details.nam_rescue
        << "\tal:i:" << details.tried_alignment
        << "\tga:i:" << details.gapped
        << "\tX0:i:" << details.best_alignments;
    sam_string.append(s.str());
}

void Sam::append_paired_details(const Details& details) {
    std::stringstream s;
    s << "\tmr:i:" << details.mate_rescue;
    sam_string.append(s.str());
}

std::string Sam::cigar_string(const Cigar& cigar) const {
    if (cigar.empty()) {
        // This case should normally not occur because
        // unmapped reads would be added with add_unmapped,
        // which hardcodes the "*"
        return "*";
    }
    if (cigar_ops == CigarOps::EQX) {
        return cigar.to_string();
    } else {
        return cigar.to_m().to_string();
    }
}

void Sam::add_unmapped(const KSeq& record, uint16_t flags) {
    if (!output_unmapped) {
        return;
    }
    assert((flags & ~(UNMAP|PAIRED|MUNMAP|READ1|READ2)) == 0);
    assert(flags & UNMAP);
    sam_string.append(strip_suffix(record.name));
    sam_string.append("\t");
    sam_string.append(std::to_string(flags));
    sam_string.append("\t*\t0\t" SAM_UNMAPPED_MAPQ_STRING "\t*\t*\t0\t0\t");
    append_seq(record.seq);
    append_qual(record.qual);
    append_rg();
    if (this->fastq_comments) {
        sam_string.append("\t");
        sam_string.append(record.comment);
    }
    sam_string.append("\n");
}

void Sam::add_unmapped_mate(const KSeq& record, uint16_t flags, const std::string& mate_reference_name, uint32_t mate_pos) {
    assert(flags & (UNMAP|PAIRED));
    sam_string.append(strip_suffix(record.name));
    sam_string.append("\t");
    sam_string.append(std::to_string(flags));
    sam_string.append("\t");
    // The SAM specification recommends: "For a unmapped paired-end or
    // mate-pair read whose mate is mapped, the unmapped read should have
    // RNAME and POS identical to its mate."
    sam_string.append(mate_reference_name);
    sam_string.append("\t");
    sam_string.append(std::to_string(mate_pos + 1));
    sam_string.append("\t" SAM_UNMAPPED_MAPQ_STRING "\t*\t");
    sam_string.append("=");
    sam_string.append("\t");
    sam_string.append(std::to_string(mate_pos + 1));
    sam_string.append("\t0\t");
    append_seq(record.seq);
    append_qual(record.qual);
    append_rg();
    if (this->fastq_comments) {
        sam_string.append("\t");
        sam_string.append(record.comment);
    }
    sam_string.append("\n");
}

void Sam::add_unmapped_pair(const KSeq& r1, const KSeq& r2) {
    add_unmapped(r1, PAIRED | UNMAP | MUNMAP | READ1);
    add_unmapped(r2, PAIRED | UNMAP | MUNMAP | READ2);
}

// Add single-end alignment
void Sam::add(
    const Alignment& alignment,
    const KSeq& record,
    const std::string& sequence_rc,
    uint8_t mapq,
    bool is_primary,
    const Details& details
) {
    assert(!alignment.is_unaligned);

    int flags = 0;
    if (!alignment.is_unaligned && alignment.is_rc) {
        flags |= REVERSE;
    }
    if (!is_primary) {
        flags |= SECONDARY;
        mapq = 255;
    }
    add_record(record.name, record.comment, flags, references.names[alignment.ref_id], alignment.ref_start, mapq, alignment.cigar, "*", -1, 0, record.seq, sequence_rc, record.qual, alignment.edit_distance, alignment.score, details);
}

// Add one individual record
void Sam::add_record(
    const std::string& query_name,
    const std::string& comment,
    uint16_t flags,
    const std::string& reference_name,
    uint32_t pos,
    uint8_t mapq,
    const Cigar& cigar,
    const std::string& mate_reference_name,
    uint32_t mate_pos,
    int32_t template_len,
    const std::string& query_sequence,
    const std::string& query_sequence_rc,
    const std::string& qual,
    int ed,
    int aln_score,
    const Details& details
) {
    sam_string.append(strip_suffix(query_name));
    sam_string.append("\t");
    sam_string.append(std::to_string(flags));
    sam_string.append("\t");
    sam_string.append(reference_name);
    sam_string.append("\t");
    sam_string.append(std::to_string(pos + 1));  // convert to 1-based coordinate
    sam_string.append("\t");
    sam_string.append(std::to_string(mapq));
    sam_string.append("\t");
    sam_string.append(cigar_string(cigar));
    sam_string.append("\t");

    sam_string.append(mate_reference_name);
    sam_string.append("\t");
    sam_string.append(std::to_string(mate_pos + 1));
    sam_string.append("\t");
    sam_string.append(std::to_string(template_len));
    sam_string.append("\t");

    if (flags & SECONDARY) {
        append_seq("");
    } else if (flags & REVERSE) {
        append_seq(query_sequence_rc);
    } else {
        append_seq(query_sequence);
    }

    if (!(flags & UNMAP)) {
        if (flags & SECONDARY) {
            append_qual("");
        } else if (flags & REVERSE) {
            auto qual_rev = qual;
            std::reverse(qual_rev.begin(), qual_rev.end());
            append_qual(qual_rev);
        } else {
            append_qual(qual);
        }
        sam_string.append("\t");
        sam_string.append("NM:i:");
        sam_string.append(std::to_string(ed));
        sam_string.append("\t");
        sam_string.append("AS:i:");
        sam_string.append(std::to_string(aln_score));
    } else {
        append_qual(qual);
    }

    if (show_details) {
        append_details(details);
        if (flags & PAIRED) {
            append_paired_details(details);
        }
    }
    append_rg();
    if (this->fastq_comments) {
        sam_string.append("\t");
        sam_string.append(comment);
    }
    sam_string.append("\n");
}

void Sam::add_pair(
    const Alignment &alignment1,
    const Alignment &alignment2,
    const KSeq& record1,
    const KSeq& record2,
    const std::string &read1_rc,
    const std::string &read2_rc,
    uint8_t mapq1,
    uint8_t mapq2,
    bool is_proper,
    bool is_primary,
    const std::array<Details, 2>& details
) {
    int f1 = PAIRED | READ1;
    int f2 = PAIRED | READ2;
    if (!is_primary) {
        f1 |= SECONDARY;
        f2 |= SECONDARY;
    }

    int template_len1 = 0;
    bool both_aligned = !alignment1.is_unaligned && !alignment2.is_unaligned;
    if (both_aligned && alignment1.ref_id == alignment2.ref_id) {
        const int dist = alignment2.ref_start - alignment1.ref_start;
        if (dist > 0) {
            template_len1 = dist + alignment2.length;
        }
        else {
            template_len1 = dist - alignment1.length;
        }
    }
    if (is_proper) {
        f1 |= PROPER_PAIR;
        f2 |= PROPER_PAIR;
    }

    std::string reference_name1;
    int pos1 = alignment1.ref_start;
    int edit_distance1 = alignment1.edit_distance;
    if (alignment1.is_unaligned) {
        f1 |= UNMAP;
        f2 |= MUNMAP;
        pos1 = -1;
        reference_name1 = "*";
    } else {
        if (alignment1.is_rc) {
            f1 |= REVERSE;
            f2 |= MREVERSE;
        }
        reference_name1 = references.names[alignment1.ref_id];
    }

    std::string reference_name2;
    int pos2 = alignment2.ref_start;
    int edit_distance2 = alignment2.edit_distance;
    if (alignment2.is_unaligned) {
        f2 |= UNMAP;
        f1 |= MUNMAP;
        pos2 = -1;
        reference_name2 = "*";
    } else {
        if (alignment2.is_rc) {
            f1 |= MREVERSE;
            f2 |= REVERSE;
        }
        reference_name2 = references.names[alignment2.ref_id];
    }

    // Reference name as used in the RNEXT field;
    // set to "=" if identical to reference_name
    std::string mate_reference_name1 = reference_name1;
    std::string mate_reference_name2 = reference_name2;
    if (
        (!alignment1.is_unaligned && !alignment2.is_unaligned && alignment1.ref_id == alignment2.ref_id)
        || (alignment1.is_unaligned != alignment2.is_unaligned)
    ) {
        mate_reference_name1 = "=";
        mate_reference_name2 = "=";
    }

    if (alignment1.is_unaligned != alignment2.is_unaligned) {
        if (alignment1.is_unaligned) {
            pos1 = pos2;
        } else {
            pos2 = pos1;
        }
    }

    if (alignment1.is_unaligned) {
        add_unmapped_mate(record1, f1, reference_name2, pos2);
    } else {
        add_record(record1.name, record1.comment, f1, reference_name1, alignment1.ref_start, mapq1, alignment1.cigar, mate_reference_name2, pos2, template_len1, record1.seq, read1_rc, record1.qual, edit_distance1, alignment1.score, details[0]);
    }
    if (alignment2.is_unaligned) {
        add_unmapped_mate(record2, f2, reference_name1, pos1);
    } else {
        add_record(record2.name, record2.comment, f2, reference_name2, alignment2.ref_start, mapq2, alignment2.cigar, mate_reference_name1, pos1, -template_len1, record2.seq, read2_rc, record2.qual, edit_distance2, alignment2.score, details[1]);
    }
}

bool is_proper_pair(const Alignment& alignment1, const Alignment& alignment2, float mu, float sigma) {
    const int dist = alignment2.ref_start - alignment1.ref_start;
    const bool same_reference = alignment1.ref_id == alignment2.ref_id;
    const bool both_aligned = same_reference && !alignment1.is_unaligned && !alignment2.is_unaligned;
    const bool r1_r2 = !alignment1.is_rc && alignment2.is_rc && dist >= 0; // r1 ---> <---- r2
    const bool r2_r1 = !alignment2.is_rc && alignment1.is_rc && dist <= 0; // r2 ---> <---- r1
    const bool rel_orientation_good = r1_r2 || r2_r1;
    const bool insert_good = std::abs(dist) <= mu + 6 * sigma;

    return both_aligned && insert_good && rel_orientation_good;
}

File: src/sam.rs
use std::fmt::{Display, Formatter};
use crate::cigar::Cigar;
use crate::details::Details;
use crate::fasta::RefSequence;

pub const PAIRED: u16 = 1;
pub const PROPER_PAIR: u16 = 2;
pub const UNMAP: u16 = 4;
pub const MUNMAP: u16 = 8;
pub const REVERSE: u16 = 0x10;
pub const MREVERSE: u16 = 0x20;
pub const READ1: u16 = 0x40;
pub const READ2: u16 = 0x80;
pub const SECONDARY: u16 = 0x100;
pub const QCFAIL: u16 = 0x200;
pub const DUP: u16 = 0x400;
pub const SUPPLEMENTARY: u16 = 0x800;

pub fn strip_suffix(s: &str) -> &str {
    if s.ends_with("/1") || s.ends_with("/2") {
        &s[..s.len() - 2]
    } else {
        s
    }
}

// TODO String and Vec<u8> fields should be references
#[derive(Default, Debug)]
pub struct SamRecord {
    pub query_name: String,
    pub flags: u16,
    pub reference_name: Option<String>,
    /// 0-based position, converted to 1-based for output
    pub pos: Option<u32>,
    pub mapq: u8,
    pub cigar: Option<Cigar>,
    pub mate_reference_name: Option<String>,
    pub mate_pos: Option<u32>,
    pub template_len: Option<i32>,
    pub query_sequence: Option<Vec<u8>>,
    pub query_qualities: Option<Vec<u8>>,
    pub edit_distance: Option<u32>,
    pub alignment_score: Option<u32>,
    pub details: Option<Details>,
    pub rg_id: Option<String>,
    /// Copied from FASTQ header
    pub extra: Option<String>,
}

impl SamRecord {
    pub fn is_secondary(&self) -> bool {
        self.flags & SECONDARY != 0
    }

    pub fn is_mapped(&self) -> bool {
        self.flags & UNMAP == 0
    }
}

impl Display for SamRecord {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        // SAM columns
        //
        // 1 QNAME query template name
        // 2 FLAG  bitwise flag
        // 3 RNAME reference sequence name
        // 4 POS   1-based leftmost mapping position (set to 0 for unmapped read w/o pos)
        // 5 MAPQ  mapping quality (255 if unavailable)
        // 6 CIGAR
        // 7 RNEXT reference name of mate/next read
        // 8 PNEXT position of mate/next read
        // 9 TLEN template length
        // 10 SEQ
        // 11 QUAL
        // 12 optional fields

        let mate_pos = match self.mate_pos {
            Some(pos) => pos + 1,
            None => 0,
        };
        let query_sequence = match &self.query_sequence {
            Some(seq) if !self.is_secondary() && !seq.is_empty() => std::str::from_utf8(seq).unwrap(),
            _ => "*",
        };
        let query_qualities = match &self.query_qualities {
            Some(query_qualities) if !self.is_secondary() && !query_qualities.is_empty() => std::str::from_utf8(query_qualities).unwrap(),
            _ => "*",
        };
        let pos = match self.pos {
            Some(pos) => pos + 1,
            None => 0,
        };
        let cigar = match &self.cigar {
            Some(cigar) => cigar.to_string(),
            None => "*".to_string(),
        };
        write!(f, "{}\t{}\t{}\t{}\t{}\t{}\t{}\t{}\t{}\t{}\t{}",
            self.query_name,
            self.flags,
            self.reference_name.as_ref().unwrap_or(&"*".to_string()),
            pos,
            self.mapq,
            cigar,
            self.mate_reference_name.as_ref().unwrap_or(&"*".to_string()),
            mate_pos,
            self.template_len.unwrap_or(0),
            query_sequence,
            query_qualities,
        )?;
        if let Some(edit_distance) = self.edit_distance {
            write!(f, "\tNM:i:{}", edit_distance)?;
        }
        if let Some(alignment_score) = self.alignment_score {
            write!(f, "\tAS:i:{}", alignment_score)?;
        }
        if let Some(details) = &self.details {
            let nr = if details.nam.nam_rescue > 0 {
                details.nam.n_rescue_nams as i64
            } else {
                -1
            };
            write!(f, "\tna:i:{}\tnr:i:{}\tal:i:{}\tga:i:{}\tX0:i:{}",
                details.nam.n_nams,
                nr,
                details.tried_alignment,
                details.gapped,
                details.best_alignments,
            )?;
            if self.flags & PAIRED != 0 {
                write!(f, "\tmr:i:{}", details.mate_rescue)?;
            }
        }
        if let Some(rg_id) = &self.rg_id {
            write!(f, "\tRG:Z:{}", rg_id)?;
        }
        if let Some(extra) = &self.extra {
            write!(f, "\t{}", extra)?;
        }
        Ok(())
    }
}

pub struct ReadGroup {
    id: String,
    fields: Vec<String>,
}

impl ReadGroup {
    pub fn new(id: &str, fields: Vec<String>) -> Self {
        ReadGroup {
            id: id.to_string(),
            fields,
        }
    }
}

pub struct SamHeader<'a> {
    references: &'a [RefSequence],
    cmd_line: Option<&'a str>,
    version: &'a str,
    read_group: Option<ReadGroup>,
}

impl<'a> SamHeader<'a> {
    pub fn new(
        references: &'a [RefSequence],
        cmd_line: Option<&'a str>,
        version: &'a str,
        read_group: Option<ReadGroup>,
    ) -> Self {
        SamHeader {
            references,
            cmd_line,
            version,
            read_group,
        }
    }
}

impl<'a> Display for SamHeader<'a> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "@HD\tVN:1.6\tSO:unsorted")?;

        for refseq in self.references {
            writeln!(f, "@SQ\tSN:{}\tLN:{}", refseq.name, refseq.sequence.len())?;
        }
        if let Some(read_group) = &self.read_group {
            write!(f, "@RG\tID:{}", read_group.id)?;
            for field in &read_group.fields {
                write!(f, "\t{}", &field)?;
            }
            f.write_str("\n")?;
        }
        if let Some(cmd_line) = self.cmd_line {
            writeln!(f, "@PG\tID:strobealign\tPN:strobealign\tVN:{}\tCL:{}", &self.version, &cmd_line)?;
        };

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use crate::sam::strip_suffix;

    #[test]
    fn test_strip_suffix() {
        assert_eq!(strip_suffix("abc"), "abc");
        assert_eq!(strip_suffix("abc/1"), "abc");
        assert_eq!(strip_suffix("abc/2"), "abc");
        assert_eq!(strip_suffix("abc/3"), "abc/3");
    }
}
